.circleci/config.yml

```diff
@@ -106,36 +69,68 @@ jobs:
 
 
 workflows:
-  version: 2
-  build-test-deploy:
+  ci:
     jobs:
-      - build:
+      - docker/publish:
+          name: Docker build test
+          before_build: &version
+            - run:
+                name: Generate build version.json
+                command: >
+                  printf 
+                  '{"commit":"%s","version":"%s", "source":"https://github.com/%s/%s","build":"%s"}\n' 
+                  "$CIRCLE_SHA1" "$CIRCLE_TAG" "$CIRCLE_PROJECT_USERNAME" 
+                  "$CIRCLE_PROJECT_REPONAME" "$CIRCLE_BUILD_URL" 
+                  > version.json
+            - run:
+                name: Output version.json
+                command: cat version.json
+          deploy: false
+          image: $CIRCLE_PROJECT_REPONAME
+          tag: ${CIRCLE_SHA1:0:9}
           filters:
             tags:
               only: /.*/
 
-      - test:
+      - unit-tests:
+          name: Unit tests
           filters:
             tags:
               only: /.*/
 
-      - test-environment:
+      - integration-tests:
+          name: Integration tests
           filters:
             tags:
               only: /.*/
 
-      - deploy:
-          requires:
-            - test
+  publish:
+    jobs:
+      - docker/publish:
+          name: Push latest
+          before_build: *version
+          docker-password: DOCKER_PASS
+          docker-username: DOCKER_USER
+          image: $DOCKERHUB_REPO
+          tag: latest
+          filters:
+            branches:
+              only: main
+
+      - docker/publish:
+          name: Push tag
+          before_build: *version
+          docker-password: DOCKER_PASS
+          docker-username: DOCKER_USER
+          image: $DOCKERHUB_REPO
+          tag: $CIRCLE_TAG
           filters:
             tags:
               only: /.*/
             branches:
-              only: main
+              ignore: /.*/
```

Comment:
what are the implications of this?

Reply:
Default behaviour is CircleCI will run workflows for all branch commits. By setting it to ignore all branches, we ensure it won't run all the time. A tag commit does not have a branch, so this workflow will run for all tags.